### PR TITLE
Use controller-runtime library

### PR DESCRIPTION
**Pull request by saheienko**
**What type of PR is this?**
dependencies update

**What this PR does / why we need it**:
Update the operator-sdk dependencies to v0.10.1.

**Does this PR change a user-facing CRD or CLI?**:
no

**Is a release note needed?**:
no

**Does this change need to be cherry-picked to a release branch?**:
no

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -1,96 +0,0 @@
...
-	if controllerInst.started {
-		return fmt.Errorf("can't register new handlers after starting controller")
-	}
-
-	objectType := gkv.String()
-	// Only add Watch if we aren't already watching for the object already.
-	// resyncPeriod will be ignored for second call if different
-	if controllerInst.handlers[objectType] == nil {
-		controllerInst.handlers[objectType] = make([]sdk.Handler, 0)
-		sdk.Watch(gkv.GroupVersion().String(), gkv.Kind, namespace, resyncPeriod, sdk.WithNumWorkers(10))
```
Comment (disrani-px) [inline pkg/controller/controller.go]: I don't see the number of workers being set for the new controllers

File: pkg/groupsnapshot/controllers/groupsnapshot.go [hunk 2]
```diff
@@ -72,110 +88,145 @@ func (m *GroupSnapshotController) Init() error {
...
-			if snapVersion.LessThan(minVersion) {
-				log.GroupSnapshotLog(groupSnapshot).Infof(
-					"Already processed groupSnapshot version (%s) higher than: %s. Skipping event.",
-					minVer, groupSnapshot.ResourceVersion)
-				return nil
+func (m *GroupSnapshotController) handle(ctx context.Context, groupSnapshot *stork_api.GroupVolumeSnapshot) error {
+	if groupSnapshot.DeletionTimestamp != nil {
+		if controllerutils.ContainsFinalizer(groupSnapshot, cleanupFinalizer) {
+			if err := m.handleDelete(groupSnapshot); err != nil {
+				logrus.Errorf("%s: cleanup: %s", reflect.TypeOf(m), err)
```
Comment (disrani-px) [inline pkg/groupsnapshot/controllers/groupsnapshot.go]: We should not remove the finalizer in case of errors during delete here.

File: pkg/snapshot/snapshot.go [hunk 3]
```diff
@@ -8,11 +8,12 @@ import (
 	snapshotvolume "github.com/kubernetes-incubator/external-storage/snapshot/pkg/volume"
 	"github.com/libopenstorage/stork/drivers/volume"
 	"github.com/libopenstorage/stork/pkg/snapshot/controllers"
-	"github.com/portworx/sched-ops/k8s/errors"
+	schederrors "github.com/portworx/sched-ops/k8s/errors"
```
Comment (disrani-px) [inline pkg/snapshot/snapshot.go]: Don't need to alias unless there is a conflict

File: pkg/snapshot/controllers/snapshotschedule.go [hunk 4]
```diff
@@ -37,92 +41,126 @@ const (
...
-	return controller.Register(
-		&schema.GroupVersionKind{
-			Group:   stork.GroupName,
-			Version: stork_api.SchemeGroupVersion.Version,
-			Kind:    reflect.TypeOf(stork_api.VolumeSnapshotSchedule{}).Name(),
-		},
-		"",
-		1*time.Minute,
-		s)
+	mgr.GetCache()
```
Comment (disrani-px) [inline pkg/snapshot/controllers/snapshotschedule.go]: Why is this required only in this controller?

File: pkg/snapshot/controllers/snapshotschedule.go [hunk 5]
```diff
@@ -8,26 +8,30 @@ import (
...
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
 const (
 	nameTimeSuffixFormat string        = "2006-01-02-150405"
 	validateCRDInterval  time.Duration = 5 * time.Second
 	validateCRDTimeout   time.Duration = 1 * time.Minute
+	resyncPeriod                       = 60 * time.Second
```
Comment (disrani-px) [inline pkg/snapshot/controllers/snapshotschedule.go]: Move to `pkg/controller`

File: pkg/controller/controller.go (hunk 1 above)
Comment (saheienko) [reply pkg/controller/controller.go]: There is no such parameter for controller-runtime manager. It start each controller in its goroutine.

File: Gopkg.toml [hunk 6]
```diff
@@ -135,3 +148,15 @@ required = [
   name = "github.com/kubernetes-sigs/aws-ebs-csi-driver"
   branch = "master"
 
+[[override]]
+  name = "github.com/libopenstorage/secrets"
+  branch = "sched-ops-fix"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: This PR requires a new sched-ops package and this existed only for this branch. Going to use master now.

File: drivers/volume/aws/aws.go [hunk 7]
```diff
@@ -21,7 +21,7 @@ import (
 	"github.com/portworx/sched-ops/k8s/core"
 	"github.com/portworx/sched-ops/k8s/storage"
 	"github.com/sirupsen/logrus"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply drivers/volume/aws/aws.go]: This was added by goimports formatting.

File: pkg/snapshot/snapshot.go (hunk 3 above)
Comment (saheienko) [reply pkg/snapshot/snapshot.go]: Update this but it's better to keep the same names over project.

File: pkg/snapshot/controllers/snapshotschedule.go (hunk 4 above)
Comment (saheienko) [reply pkg/snapshot/controllers/snapshotschedule.go]: There is no need for this here, removing.

File: Gopkg.toml [hunk 8]
```diff
@@ -79,36 +37,44 @@ required = [
...
-  version = "kubernetes-1.16.6"
+[[override]]
+  name = "github.com/operator-framework/operator-sdk"
+  version = "=v0.10.1" 
 
 [[override]]
-  name = "github.com/docker/distribution"
-  revision = "edc3ab29cdff8694dd6feb85cfeb4b5f1b38ed9c"
+  name = "k8s.io/kube-openapi"
+  revision = "0cf8f7e6ed1d2e3d47d02e3b6e559369af24d803"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: There are no releases for it - https://github.com/kubernetes/kube-openapi/releases

File: Gopkg.toml [hunk 9]
```diff
@@ -79,36 +37,44 @@ required = [
...
-  name = "github.com/docker/docker"
-  revision = "4f3616fb1c112e206b88cb7a9922bf49067a7756"
+  name = "github.com/go-openapi/spec"
+  branch = "master"
 
 [[override]]
-  name = "github.com/russross/blackfriday"
-  revision = "300106c228d52c8941d4b3de6054a6062a86dda3"
+  name = "sigs.k8s.io/controller-tools"
+  revision = "9d55346c2bde73fb3326ac22eac2e5210a730207"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: Updated

File: Gopkg.toml [hunk 10]
```diff
@@ -135,3 +148,15 @@ required = [
...
+  name = "github.com/libopenstorage/secrets"
+  branch = "sched-ops-fix"
+
+[[override]]
+  name = "github.com/openshift/api"
+  revision = "ecbc01928def629f4422dcbcc3362efddfc513ee" # bump to kube 1.16
+
+[[override]]
+  name = "github.com/openshift/client-go"
+  revision = "3b6373338c9bb11fbcf329c25e81ae7e9a1ed8c6" # bump to 1.16
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: There is are no releases which contains this commit.

File: Gopkg.toml [hunk 11]
```diff
@@ -119,6 +85,53 @@ required = [
...
+  name = "github.com/docker/distribution"
+  revision = "edc3ab29cdff8694dd6feb85cfeb4b5f1b38ed9c"
+
+[[override]]
+  name = "github.com/docker/docker"
+  revision = "4f3616fb1c112e206b88cb7a9922bf49067a7756"
+
+[[override]]
+  name = "github.com/russross/blackfriday"
+  revision = "300106c228d52c8941d4b3de6054a6062a86dda3"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: Updated.

File: Gopkg.toml [hunk 12]
```diff
@@ -119,6 +85,53 @@ required = [
...
+  name = "github.com/skyrings/skyring-common"
+  branch = "master"
+
+[[override]]
+  name = "github.com/docker/distribution"
+  revision = "edc3ab29cdff8694dd6feb85cfeb4b5f1b38ed9c"
+
+[[override]]
+  name = "github.com/docker/docker"
+  revision = "4f3616fb1c112e206b88cb7a9922bf49067a7756"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: Updated.

File: Gopkg.toml [hunk 13]
```diff
@@ -119,6 +85,53 @@ required = [
...
+  name = "github.com/heptio/ark"
+  version = "v1.0.0"
+
+[[constraint]]
+  name = "github.com/skyrings/skyring-common"
+  branch = "master"
+
+[[override]]
+  name = "github.com/docker/distribution"
+  revision = "edc3ab29cdff8694dd6feb85cfeb4b5f1b38ed9c"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: Updated.

File: pkg/snapshot/controllers/snapshotschedule.go (hunk 5 above)
Comment (saheienko) [reply pkg/snapshot/controllers/snapshotschedule.go]: resyncPeriod specific for each controller. Why should we do this??

File: pkg/clusterdomains/controllers/clusterdomainstatus.go [hunk 14]
```diff
@@ -68,47 +80,71 @@ func (c *ClusterDomainsStatusController) Init() error {
...
-		if updated {
-			apiClusterDomainsStatus.Status = *currentClusterDomains
-			if err := sdk.Update(apiClusterDomainsStatus); err != nil {
-				return err
-			}
+	}
+	if updated {
+		apiClusterDomainsStatus.Status = *currentClusterDomains
+		if err := c.client.Update(context.TODO(), apiClusterDomainsStatus); err != nil {
+			return err
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/clusterdomains/controllers/clusterdomainstatus.go]: Updated.

File: pkg/clusterdomains/controllers/clusterdomainstatus.go [hunk 15]
```diff
@@ -37,29 +39,39 @@ var (
...
+		Recorder: r,
+	}
+}
+
 // ClusterDomainsStatusController clusterdomainsstatus controller
 type ClusterDomainsStatusController struct {
+	client runtimeclient.Client
+
 	Driver   volume.Driver
 	Recorder record.EventRecorder
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/clusterdomains/controllers/clusterdomainstatus.go]: Updated.

File: pkg/applicationmanager/controllers/applicationrestore.go [hunk 16]
```diff
@@ -6,42 +6,59 @@ import (
...
+		ResourceCollector: rc,
+	}
+}
+
 // ApplicationRestoreController reconciles applicationrestore objects
 type ApplicationRestoreController struct {
+	client runtimeclient.Client
+
 	Recorder              record.EventRecorder
 	ResourceCollector     resourcecollector.ResourceCollector
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/applicationmanager/controllers/applicationrestore.go]: Updated.

File: pkg/groupsnapshot/controllers/groupsnapshot.go [hunk 17]
```diff
@@ -46,6 +49,8 @@ const (
 	volumeSnapshotInitialDelay = 2 * time.Second
 	volumeSnapshotFactor       = 1
 	volumeSnapshotSteps        = 60
+
+	cleanupFinalizer = "cleanup"
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/groupsnapshot/controllers/groupsnapshot.go]: Updated.

File: pkg/applicationmanager/controllers/applicationclone.go [hunk 18]
```diff
@@ -23,19 +22,36 @@ import (
...
+	}
+}
+
 // ApplicationCloneController reconciles applicationclone objects
 type ApplicationCloneController struct {
+	client runtimeclient.Client
+
 	Driver            volume.Driver
 	Recorder          record.EventRecorder
 	ResourceCollector resourcecollector.ResourceCollector
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/applicationmanager/controllers/applicationclone.go]: Updated.

File: pkg/applicationmanager/controllers/applicationbackupschedule.go [hunk 19]
```diff
@@ -35,92 +38,124 @@ const (
...
-func (s *ApplicationBackupScheduleController) Handle(ctx context.Context, event sdk.Event) error {
-	switch o := event.Object.(type) {
-	case *stork_api.ApplicationBackupSchedule:
-		backupSchedule := o
-		// Nothing to do for delete
-		if event.Deleted {
-			return nil
+// Reconcile updates for ApplicationBackupSchedule objects.
+func (s *ApplicationBackupScheduleController) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logrus.Debugf("Reconciling ApplicationBackupSchedule %s/%s", request.Namespace, request.Name)
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/applicationmanager/controllers/applicationbackupschedule.go]: Updated.

File: cmd/stork/stork.go [hunk 20]
```diff
@@ -198,25 +199,32 @@ func run(c *cli.Context) {
...
 
+	// Create operator-sdk manager that will manage all controllers.
+	mgr, err := manager.New(config, manager.Options{})
+	if err != nil {
+		log.Fatalf("Setup controller manager: %v", err)
+	}
+
+	// Setup scheme for all stork resources
+	if err := apis.AddToScheme(mgr.GetScheme()); err != nil {
+		log.Fatalf("Setup scheme for stork resources: %v", err)
```
[the remark this answers is not in the document]
Comment (saheienko) [reply cmd/stork/stork.go]: Updated.

File: Gopkg.toml [hunk 21]
```diff
@@ -135,3 +148,15 @@ required = [
   name = "github.com/kubernetes-sigs/aws-ebs-csi-driver"
   branch = "master"
 
+[[override]]
+  name = "github.com/libopenstorage/secrets"
+  branch = "sched-ops-fix"
+
+[[override]]
+  name = "github.com/openshift/api"
+  revision = "ecbc01928def629f4422dcbcc3362efddfc513ee" # bump to kube 1.16
```
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: Using `release-4.4` branch.

File: Gopkg.toml (hunk 10 above)
[the remark this answers is not in the document]
Comment (saheienko) [reply Gopkg.toml]: Using `release-4.4` branch.

File: pkg/controller/controller.go (hunk 1 above)
Comment (saheienko) [reply pkg/controller/controller.go]: Add `MaxConcurrentReconciles` for controllers.

File: pkg/snapshot/controllers/snapshotschedule.go (hunk 5 above)
Comment (saheienko) [reply pkg/snapshot/controllers/snapshotschedule.go]: Updated.

File: pkg/applicationmanager/controllers/applicationclone.go [hunk 22]
```diff
@@ -126,117 +141,156 @@ func (a *ApplicationCloneController) verifyNamespaces(clone *stork_api.Applicati
...
+func (a *ApplicationCloneController) handle(ctx context.Context, clone *stork_api.ApplicationClone) error {
+	if clone.DeletionTimestamp != nil {
+		if controllerutils.ContainsFinalizer(clone, cleanupFinalizer) {
+			if err := a.deleteClone(clone); err != nil {
+				logrus.Errorf("%s: cleanup: %s", reflect.TypeOf(a), err)
+			}
+		}
+
+		if clone.GetFinalizers() != nil {
+			clone.SetFinalizers(nil)
```
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/applicationmanager/controllers/applicationclone.go]: Updated.

File: pkg/controllerutils/finalizers.go [hunk 23]
```diff
@@ -0,0 +1,45 @@
+package controllerutils
```
Comment (disrani-px) [inline pkg/controllerutils/finalizers.go]: Don't need a separate package for this, move this to the `controllers` package

File: pkg/applicationmanager/controllers/applicationrestore.go [hunk 24]
```diff
@@ -10,38 +10,52 @@ import (
...
+}
+
 // ApplicationRestoreController reconciles applicationrestore objects
 type ApplicationRestoreController struct {
-	Recorder              record.EventRecorder
-	ResourceCollector     resourcecollector.ResourceCollector
+	client runtimeclient.Client
+
+	recorder              record.EventRecorder
+	rssCollector          resourcecollector.ResourceCollector
```
Comment (disrani-px) [inline pkg/applicationmanager/controllers/applicationrestore.go]:
```suggestion
	resourceCollector          resourcecollector.ResourceCollector
```

File: pkg/controllerutils/finalizers.go [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+		meta.SetFinalizers(append(meta.GetFinalizers(), finalizer))
+	}
+}
+
+// RemoveFinalizer removes a finalizer if it exists.
+func RemoveFinalizer(meta metav1.Object, finalizer string) {
+	if meta == nil {
+		return
+	}
+	if !ContainsFinalizer(meta, finalizer) {
```
Comment (disrani-px) [inline pkg/controllerutils/finalizers.go]: Don't really need this, you'll go through the same finalizers here and then below again

Comment (saheienko) [reply pkg/controllerutils/finalizers.go]: Updated.

File: pkg/applicationmanager/controllers/applicationrestore.go (hunk 24 above)
Comment (saheienko) [reply pkg/applicationmanager/controllers/applicationrestore.go]: Updated.

File: pkg/controllerutils/finalizers.go (hunk 23 above)
Comment (saheienko) [reply pkg/controllerutils/finalizers.go]: Updated.
